Damage decals showing the unit damage level on icons

SJuliez opened the pull request:
Resolves #1910 

Adds damage decals to units based on entity.getDamageLevel(), showing different decals for light, moderate, heavy and crippled. Since this is pixel-added anyway, we can use any calculation for it; currently, it uses some sort of overlay mode with the example decals provided by saxarba. It currently looks a little dark-ish, I'll maybe have to change the calculation.

I found no game event that will trigger whenever a unit receives damage, not even gameEntityChanged seems to do that, so currently the update of the unit icons happens at each phase change for units that received damage this turn and have at least light damage. 

Since it will be difficult to distinguish the various damage levels, especially on small units, I suggest using a slight smoke effect for heavy dmg and a more dense one for crippled, preferably starting somewhere near the center.

Notes to self:
- [x] loading and adding decals
- [x] GUI Pref setting
- [x] actual decals (have two)
- [x] test multihex units, add decal to damaged hex only
- [x] use multiple different decals for each level? Requires more work on loading/storing/choosing them

![image](https://user-images.githubusercontent.com/17069663/85770040-d0e62d80-b71a-11ea-80b8-c53e6e037fbf.png)

NickAragua commented:
I think updating damage levels after phase change is fine. 

Agreed about differentiating damage levels with smoke or whatever other visible indicator. 

Not sure about needing a GUI pref, but it probably won't hurt (let's make it on by default though).

It would be pretty neat if we could just apply the damage decals to the damaged "sides" only of multi-hex units.

HammerGS commented:
This fits firmly into the run before we walk section. What about something like a decal for a unit hit with inferno's?

Saxarba1 commented:
I did a first version of a smoke and a smoke + fire decals that could be applied over the moderate damage to indicate heavy and crippled respectively. Only tested with the locust and behemoth graphics. They look a bit over the top in the locust and a bit scarce in the behemoth, so I guess they should work fine for all sizes.
![CrippledDamage](https://user-images.githubusercontent.com/65286217/85840103-35959c80-b79c-11ea-9b77-7700fe1068cf.png)
![HeavyDamage](https://user-images.githubusercontent.com/65286217/85840106-362e3300-b79c-11ea-9182-a6c5ff0fe9c6.png)
![DamageTest](https://user-images.githubusercontent.com/65286217/85840118-3cbcaa80-b79c-11ea-9bb1-e1b81b6e74d7.png)

NickAragua commented:
Just want to say that looks freaking awesome.

HammerGS commented:
Those look fantastic.  Have a couple of suggestions.  The decals with smoke and fire look a little linear maybe stagger them a little like a scalene triangle.  The books describe autocannons are pock marked (which you've got) but lasers and PPC's leave scars. 

But I'll echo NickArgua - freaking awesome and can't wait to see this come together.

Saxarba1 commented:
Yes, good idea. I'll move the central smoke column/fire a bit up so they are not all aligned horizontally.

Saxarba1 commented:
Also, I'll try replacing some of the pock marks with linear scars.

SJuliez commented:
Thanks @Saxarba1 !
I divided the smoke overlays. Units over 70 tons get three smokestacks, over 40 tons get 2 and the smalls get only 1.
The IMMOBILE text kinda gets in the way. I'm tempted to move that to the unit tag. I should probably exclude infantry ;)
![Smoke1](https://user-images.githubusercontent.com/17069663/85881450-ea03e280-b7dd-11ea-9149-e69d74d7e7d3.png)
![Smoke2](https://user-images.githubusercontent.com/17069663/85881453-ebcda600-b7dd-11ea-97cd-e85ec4582667.png)
![Smoke3](https://user-images.githubusercontent.com/17069663/85881461-ef612d00-b7dd-11ea-94c1-25f5ca1f94b8.png)
![SmokeFire1](https://user-images.githubusercontent.com/17069663/85881469-f12af080-b7dd-11ea-82ac-da9af088c5c1.png)
![SmokeFire2](https://user-images.githubusercontent.com/17069663/85881475-f25c1d80-b7dd-11ea-9e31-4e2a4f204550.png)
![SmokeFire3](https://user-images.githubusercontent.com/17069663/85881478-f38d4a80-b7dd-11ea-9ed2-ecb99dbf2519.png)







![image](https://user-images.githubusercontent.com/17069663/85881334-bde86180-b7dd-11ea-9fa2-503353dae066.png)

NickAragua commented:
I like the idea of moving the immobile label (for consistency, I'd almost say we should move all such labels to the unit tag).

HammerGS commented:
Having played around in the branch, this is gonna be awesome. I also agree the "Immobile" type tags got to move.

SJuliez commented:
Ok, I'll take a loot at the tags, but I'll keep it separate from this one.

SJuliez commented:
Some more changes: 
- made slim units get only 2 smokes/fires at most (tanks and Aero)
- made very slim units get only one fire/smoke (VTOL)
- only scars in the lobby
- no damage on the icons in the unit overview (they are smaller and too easily obscured)
- and refactored the tilemanager by extracting the EntityImage internal class and put tileset things into its own package

Should give this a test IMO. Possible things to add would be:
- multiple different images for the same damage (smoke and fire mostly, with the scars it's not very visible that it's always the same image)
- inferno hits (already have nice images from Saxarba, these could also be applied as a heat indicator)
- add scars to wrecks; let wrecks burn for a round, then smoke for another round...

HammerGS commented:
I know Saxarba was going to work on updated decals. Should we hold off until we have those?

NickAragua commented on megamek/src/megamek/client/ui/swing/tileset/EntityImage.java:
```diff
@@ -0,0 +1,405 @@
+        isInfantry = entity instanceof Infantry;
+        isSecondaryPos = secondaryPos != 0 && secondaryPos != -1;
+        isPreview = preview;
+        isSlim = entity instanceof Tank || entity instanceof Aero;
```
Is there a way of determining this based on the unit image (without any too crazy computations)? e.g. checking width/height?  A Demolisher tank might warrant three smoke stacks, while a Maultier APC only really needs one.

SJuliez replied:
Possible by checking the pixels in the center row. I just think that it wouldn't make a big difference; two or three smoke stacks on a unit that could hold three but burns alright with two - in other words, all the pixel-counting effort wouldn't really show in the result. I wouldnt really want to do that to increase the number of smoke stacks. If on the other hand many units showed smoke stacks where there shouldnt be any and a weight test doesnt help, then that's be a good solution.

NickAragua replied:
I'm not aware of too many skinny 100-ton tanks. I guess the Long Tom + Trailer currently? Probably let's leave it for a later PR.

NickAragua commented on megamek/src/megamek/client/ui/swing/tileset/TilesetManager.java:
```diff
@@ -0,0 +1,556 @@
+ *
+ * @author Ben
+ */
+public class TilesetManager implements IPreferenceChangeListener, ITilesetManager {
```
Did you just move this file, or did you make changes to it? If the latter, could you point to where the changes are?

SJuliez replied:
I'm really a git noob. Does it show everything as changed because I moved it?
Well in reality, I changed almost everything because there was so much redundant and (to me) unorganized code. I also moved out EntityImage to its own class and removed the default hex Tileset that was in there as a 1000-line long list of println (instead that default is now present as a simple text file in case something with the actual tileset files goes wrong... and I guess there's a few other files that can crash MM when they're not present so why worry too much about this one).

NickAragua replied:
Ok, I'll review it as "fresh".

NickAragua commented on megamek/src/megamek/client/ui/swing/tileset/TilesetManager.java:
```diff
@@ -0,0 +1,556 @@
+                hexTileset = hts;
+                boardview.clearHexImageCache();
+            } catch (IOException ex) {
+                return;
```
Is there a chance that something meaningful could happen here? If so, let's log it.

NickAragua commented on megamek/src/megamek/client/ui/swing/tileset/TilesetManager.java:
```diff
@@ -0,0 +1,556 @@
+            if (new MegaMekFile(Configuration.hexesDir(), FILENAME_DEFAULT_HEX_SET).getFile().exists()){
+                hexTileset.loadFromFile(FILENAME_DEFAULT_HEX_SET);
+            } else {
+                System.err.println("Critical Error loading default tileset!");
```
Do we want to avoid system.out.println and instead log the error?

NickAragua commented on megamek/src/megamek/client/ui/swing/tileset/TilesetManager.java:
```diff
@@ -0,0 +1,556 @@
+    public Image wreckMarkerFor(Entity entity, int secondaryPos) {
+        EntityImage entityImage = getFromCache(entity, secondaryPos);
+        if (entityImage == null) {
+            System.out.println("Unable to load wreckMarker image for entity: " + entity.getShortNameRaw()); //$NON-NLS-1$
```
Another system.out.println which will not be seen by anyone.

NickAragua commented on megamek/src/megamek/client/ui/swing/tileset/TilesetManager.java:
```diff
@@ -0,0 +1,556 @@
+    public Image imageFor(Entity entity, int facing, int secondaryPos) {
+        EntityImage entityImage = getFromCache(entity, secondaryPos);
+        if (entityImage == null) {
+            System.err.println("Unable to load image for entity: " + entity.getShortNameRaw()); //$NON-NLS-1$
```
I'm going to stop calling out system.println and just say that it would be useful to go through this file and replace them with calls to actual logging.

Saxarba1 commented:
Here are some reworked damage decals. The light and moderate decals have laser scars and fill the whole hex now, so in multi-hex units they don't appear as isolated in each hex. Also I moved the smoke and fire so they are not aligned horizontally in the cases when the three smoke stacks are used. There's also a big smoke and fire version for the central hex of multi-hex units.
![LightDamage](https://user-images.githubusercontent.com/65286217/86588708-840d1d00-bf8c-11ea-9c14-0ef9ca13a999.png)
![ModerateDamage](https://user-images.githubusercontent.com/65286217/86588710-84a5b380-bf8c-11ea-9fa4-b9d14814adbf.png)
![CentralMultiHexCrippled](https://user-images.githubusercontent.com/65286217/86588711-853e4a00-bf8c-11ea-95ec-1f69f9e7e50a.png)
![CentralMultiHexHeavy](https://user-images.githubusercontent.com/65286217/86588713-853e4a00-bf8c-11ea-9375-352c9e011d42.png)
![CrippledDamage](https://user-images.githubusercontent.com/65286217/86588715-85d6e080-bf8c-11ea-8110-474a55a16048.png)
![HeavyDamage](https://user-images.githubusercontent.com/65286217/86588716-85d6e080-bf8c-11ea-8007-d60d9a851d7d.png)

HammerGS commented:
@Saxarba1 Those are fantastic.

SJuliez commented:
@Saxarba1 Thanks!
Which blending mode were you using when painting these?
![image](https://user-images.githubusercontent.com/17069663/86803610-703de580-c076-11ea-8d9d-0deb881790b4.png)
When just overlaying it I find the laser scars don't look like scars. Using "multiply" in a gfx program works better for the scars but makes the shot marks much darker:
![image](https://user-images.githubusercontent.com/17069663/86803965-c6ab2400-c076-11ea-93b3-c869ab0a8acd.png)
Also, I think the scars are slightly visible outside the unit in the drop shadow but I can deal with that.

Saxarba1 commented:
I was using a normal overlay. The yellow/orange color was trying to achieve a residual glow in the scars. I'll make a version with the scars using the same neutral colors as the shot marks and making the scars thinner, to see if they look better.

Saxarba1 commented:
@SJuliez See if these look better. I did a couple of versions of each in case you want to add some variation. The scars repetition will probably be more noticeable than the small shots alone.
![Light1](https://user-images.githubusercontent.com/65286217/87018576-beb3c700-c1d1-11ea-8b05-e9932a588006.png)
![Light2](https://user-images.githubusercontent.com/65286217/87018581-bfe4f400-c1d1-11ea-9d7e-a52b4aeddb11.png)
![Moderate1](https://user-images.githubusercontent.com/65286217/87018583-bfe4f400-c1d1-11ea-8123-64bf50581cde.png)
![Moderate2](https://user-images.githubusercontent.com/65286217/87018586-c07d8a80-c1d1-11ea-8074-54f8f5f557e3.png)

SJuliez commented:
Using the MMLogger now. Also removed the decals from drop shadows and added support for the bigger multi-hex image for dropships.

SJuliez commented:
Not quite finished... Need to add the changed small smoke/fire, decide if its worth adding support for multiple images

HammerGS commented:
You had some feedback tossed up in Slack on the test images you posted.

SJuliez commented:
Saw it, already changed the scar images used to the latest version from Saxarba. But he provided two of each and I could make up a few different smoke/fire images from the ones I now have. But it's not a one-liner code addition to load and choose between an arbitrary number of images. Will have to sleep on it (and get through a work monday).

SJuliez commented:
OK, now it supports multiple images for everything. The images are organized into directories, with Crippled/ being for the scars of crippled units and smoke1/ for the smoke images of very slim/light units and so on. The names of the images don't matter and we can put in as many as we like. I created a few from Saxarba's images.

From my side this is ready for review. @NickAragua maybe you want to take a look at the changes again. The DirectoryItems class was very practical for this, but I noticed that the ImageFileFactory (=FilenameFilter) that is often used with it scales all images to hex size (a little evil with such a generic name). I didnt want to change the places where it is used so I renamed it to ScaledImageFileFactory and added ImageFileFactory which now doesnt scale images.

![image](https://user-images.githubusercontent.com/17069663/87966163-27961b80-cabd-11ea-9f40-9ca7b205d92a.png)

NickAragua commented on megamek/src/megamek/client/ui/swing/util/ScaledImageFileFactory.java:
```diff
@@ -0,0 +1,215 @@
+                if (null == image) {
+
+                    // Get ready to read from the item.
+                    InputStream in = new BufferedInputStream(zipFile
```
According to LGTM (and it seems to be right), you should close this input stream regardless of whether the call succeeds or fails.
